### PR TITLE
Sagemaker Endpoint Resource 

### LINKED ISSUE #1029: Request for `aws_sagemaker_endpoint` resource
### What resource do you need?

Terraform Resource Name:  `aws_sagemaker_endpoint`

### What is your use case?

We want to deploy SageMaker models with Crossplane, We currently can configure endpoints with the `aws_sagemaker_endpoint_configuration` resource but can't actually create an endpoint without the SageMaker Endpoint resource.

### PR DESCRIPTION
**Pull request by blakeromano**
### Description of your changes

Adds Sagemaker Endpoint resource.

Fixes #1029 

I have:

- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

This requires an image that can be created/deleted only from the AWS CLI. In this case I use a public image from ECR of `683313688378.dkr.ecr.us-east-1.amazonaws.com/sagemaker-scikit-learn:0.23-1-cpu-py3` found here: https://docs.aws.amazon.com/sagemaker/latest/dg-ecr-paths/ecr-us-east-2.html#sklearn-us-east-2.title
I update the Model resource with the full image reference for the key: `spec.forProvider.primaryContainer[0]`

I see the following output once I apply that and let resources create:

Screenshot 2023-12-26 at 11 40 23 AM

### CODE REVIEW

**Review by mbbush — COMMENTED** [review]
Thanks for the contribution! It sounds like you need a bit of help getting your local dev environment set up properly. The command to use to run codegen is simply `make generate`. You should also test your example using `make e2e UPTEST_EXAMPLE_LIST=config/examples/sagemaker/endpoint.yaml`. You'll need to install kind, kubectl, and a few other tools, and configure aws credentials in an env var as explained in a comment in the Makefile. Yes, you can make an env var with newlines in it.

File: examples/sagemaker/endpoint.yaml [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  labels:
+    testing.upbound.io/example-name: example
+  name: example
+spec:
+  forProvider:
+    executionRoleArnSelector:
+      matchLabels:
+        testing.upbound.io/example-name: example
+    primaryContainer:
+      - image: ${data.aws_account_id}.dkr.ecr.us-east-1.amazonaws.com/sagemaker-scikit-learn:latest
```
Comment (mbbush) [inline examples/sagemaker/endpoint.yaml:56]: Is this the only reason for the manual intervention annotation? I'm not very familiar with Sagemaker, but couldn't you just define the endpoint and EndpointConfiguration in here, and not have a model?

Comment (blakeromano) [reply examples/sagemaker/endpoint.yaml:57]: You need to have a model for an EndpointConfiguration and need an EndpointConfiguration for an Endpoint hence the need for the label

File: config/sagemaker/config.go [hunk 2]
```diff
@@ -14,4 +14,10 @@ func Configure(p *config.Provider) {
 	p.AddResourceConfigurator("aws_sagemaker_device_fleet", func(r *config.Resource) {
 		r.Path = "devicefleet"
 	})
+	p.AddResourceConfigurator("aws_sagemaker_endpoint", func(r *config.Resource) {
+		r.References["endpoint_config_name"] = config.Reference{
+			Type: "EndpointConfiguration",
```
Comment (turkenf) [inline config/sagemaker/config.go:19]: Could you please use the following configuration which is safer?

```suggestion
			TerraformName: "aws_sagemaker_endpoint_configuration",
```

Comment (turkenf) [inline config/sagemaker/config.go:19]: Could you please use the following configuration which is safer?

```suggestion
			TerraformName: "aws_sagemaker_endpoint_configuration",
```

File: examples/sagemaker/endpoint.yaml [hunk 3]
```diff
@@ -0,0 +1,100 @@
+---
```
Comment (turkenf) [inline examples/sagemaker/endpoint.yaml:1]: Could you please remove this separator?

File: examples/sagemaker/endpoint.yaml [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+apiVersion: sagemaker.aws.upbound.io/v1beta1
+kind: Endpoint
+metadata:
+  annotations:
+    meta.upbound.io/example-id: sagemaker/v1beta1/endpoint
+  labels:
+    testing.upbound.io/example-name: e
+    upjet.upbound.io/manual-intervention: "It requires an image which can be create/deleted only using AWS CLI"
```
Comment (turkenf) [inline examples/sagemaker/endpoint.yaml:9]: We define manual interventions in the annotations, please remove from here.

File: examples/sagemaker/endpoint.yaml [hunk 5]
```diff
@@ -0,0 +1,100 @@
+---
+apiVersion: sagemaker.aws.upbound.io/v1beta1
+kind: Endpoint
+metadata:
+  annotations:
+    meta.upbound.io/example-id: sagemaker/v1beta1/endpoint
```
Comment (turkenf) [inline examples/sagemaker/endpoint.yaml:6]:
```suggestion
    meta.upbound.io/example-id: sagemaker/v1beta1/endpoint
    upjet.upbound.io/manual-intervention: "This resource is skipping because the EndpointConfiguration resource on which this resource depends was skipped."
```

File: examples/sagemaker/endpoint.yaml [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        testing.upbound.io/example-name: ec
+    region: us-east-1
+    tags:
+      Name: foo
+---
+apiVersion: sagemaker.aws.upbound.io/v1beta1
+kind: EndpointConfiguration
+metadata:
+  annotations:
+    meta.upbound.io/example-id: sagemaker/v1beta1/endpointconfiguration
```
Comment (turkenf) [inline examples/sagemaker/endpoint.yaml:24]: Could you please fix example-id?

```suggestion
    meta.upbound.io/example-id: sagemaker/v1beta1/endpoint
```

File: examples/sagemaker/endpoint.yaml [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+      variantName: variant-1
+    region: us-east-1
+    tags:
+      Name: foo
+---
+apiVersion: sagemaker.aws.upbound.io/v1beta1
+kind: Model
+metadata:
+  annotations:
+    meta.upbound.io/example-id: sagemaker/v1beta1/endpointconfiguration
```
Comment (turkenf) [inline examples/sagemaker/endpoint.yaml:46]:
```suggestion
    meta.upbound.io/example-id: sagemaker/v1beta1/endpoint
```

File: examples/sagemaker/endpoint.yaml [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        testing.upbound.io/example-name: example
+    primaryContainer:
+      - image: ${data.aws_account_id}.dkr.ecr.us-east-1.amazonaws.com/sagemaker-scikit-learn:latest
+    region: us-east-1
+---
+apiVersion: iam.aws.upbound.io/v1beta1
+kind: Role
+metadata:
+  annotations:
+    meta.upbound.io/example-id: sagemaker/v1beta1/endpointconfiguration
```
Comment (turkenf) [inline examples/sagemaker/endpoint.yaml:64]:
```suggestion
    meta.upbound.io/example-id: sagemaker/v1beta1/endpoint
```

File: examples/sagemaker/endpoint.yaml [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+            "Sid": ""
+          }
+        ]
+      }
+---
+apiVersion: iam.aws.upbound.io/v1beta1
+kind: RolePolicyAttachment
+metadata:
+  annotations:
+    meta.upbound.io/example-id: sagemaker/v1beta1/endpointconfiguration
```
Comment (turkenf) [inline examples/sagemaker/endpoint.yaml:90]:
```suggestion
    meta.upbound.io/example-id: sagemaker/v1beta1/endpoint
```

**Review by turkenf — COMMENTED** [review]
Thank you for your efforts in this PR @blakeromano, I left a few comments for you to consider.
Please fill in how you tested the resource after all corrections as in this [PR](https://github.com/upbound/provider-aws/pull/984#issue-2010185722).

**Comment by blakeromano** [comment]
I believe I have fixed all the PR comments and have updated my comment with a way I was able to validate endpoint creation being successful with Crossplane locally. @turkenf

File: examples/sagemaker/endpoint.yaml [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+  labels:
+    testing.upbound.io/example-name: example
+  name: example
+spec:
+  forProvider:
+    executionRoleArnSelector:
+      matchLabels:
+        testing.upbound.io/example-name: example
+    primaryContainer:
+      - image: 683313688378.dkr.ecr.us-east-1.amazonaws.com/sagemaker-scikit-learn:0.23-1-cpu-py3
```
Comment (turkenf) [inline examples/sagemaker/endpoint.yaml:56]:
```suggestion
      - image: ${data.aws_account_id}.dkr.ecr.us-east-1.amazonaws.com/sagemaker-scikit-learn:latest
```

**Review by turkenf — APPROVED** [review]
Many thanks for your contribution @blakeromano, I just left a small comment that you should do before merging. Since these examples will appear on the marketplace, it would be better to remove your account ID as I mentioned in the comment below.

File: examples/sagemaker/endpoint.yaml (hunk 10 above)
Comment (blakeromano) [reply examples/sagemaker/endpoint.yaml]: Totally makes sense. I wasn't overly concerned just for the fact that this is a public image but I agree with your comment here.
